crates/red_knot_python_semantic/src/core_stdlib_modules.rs

```diff
@@ -0,0 +1,38 @@
+use crate::module_name::ModuleName;
+use crate::module_resolver::resolve_module;
+use crate::semantic_index::global_scope;
+use crate::semantic_index::symbol::ScopeId;
+use crate::Db;
+
+/// Salsa query to get the builtins scope.
+///
+/// Can return None if a custom typeshed is used that is missing `builtins.pyi`.
+#[salsa::tracked]
+pub(crate) fn builtins_scope(db: &dyn Db) -> Option<ScopeId<'_>> {
+    let builtins_name =
+        ModuleName::new_static("builtins").expect("Expected 'builtins' to be a valid module name");
+    let builtins_file = resolve_module(db, builtins_name)?.file();
+    Some(global_scope(db, builtins_file))
+}
+
+/// Salsa query to get the scope for the `types` module.
+///
+/// Can return None if a custom typeshed is used that is missing `types.pyi`.
+#[salsa::tracked]
+pub(crate) fn types_scope(db: &dyn Db) -> Option<ScopeId<'_>> {
+    let types_module_name =
+        ModuleName::new_static("types").expect("Expected 'types' to be a valid module name");
+    let types_file = resolve_module(db, types_module_name)?.file();
+    Some(global_scope(db, types_file))
+}
+
+/// Salsa query to get the scope for the `_typeshed` module.
+///
+/// Can return None if a custom typeshed is used that is missing a `_typeshed` directory.
+#[salsa::tracked]
```

Comment:
It's probably unrelated to this PR but I still wonder if it is worth having this as a salsa tracked considering that both `resolve_module` and `global_scope` are salsa ingredients.

Reply:
It's possible to get rid of this and the other two as standalone functions, and just have the `Stdlib` enum:

```rs
/// Enumeration of various core stdlib modules
/// which we need frequent access to during semantic analysis
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Stdlib {
    Builtins,
    Types,
    Typeshed,
}

impl Stdlib {
    pub(crate) fn module_name(self) -> &'static str {
        match self {
            Self::Builtins => "builtins",
            Self::Types => "types",
            Self::Typeshed => "_typeshed",
        }
    }

    /// Retrieve the global scope of the given module.
    ///
    /// Returns `None` if the given module isn't available for some reason.
    pub(crate) fn global_scope(self, db: &dyn Db) -> Option<ScopeId<'_>> {
        let module_name = self.module_name();
        let module_name = ModuleName::new_static(module_name)
            .unwrap_or_else(|| panic!("Expected '{module_name}' to be a valid module name"));
        let builtins_file = resolve_module(db, module_name)?.file();
        Some(global_scope(db, builtins_file))
    }

    /// Look up a symbol in the global scope of the given module.
    ///
    /// Returns `Unbound` if the given module isn't available for some reason.
    pub(crate) fn global_symbol<'db>(self, db: &'db dyn Db, name: &str) -> Type<'db> {
        self.global_scope(db)
            .map(|globals| symbol_ty_by_name(db, globals, name))
            .unwrap_or(Type::Unbound)
    }
}
```

All uses of `builtins_symbol_ty_by_name()` would be replaced with `Stdlib::Builtins.global_symbol()`.

I like that design more in lots of ways, but it's also a much bigger diff, and it means that the function to get the builtins scope can't be Salsa-tracked. Overall I'm definitely open to it but I'd rather do it as a followup, as I'm uncertain what impact it would have on performance.

Reply:
I think it's probably not worth having these as tracked functions, but fine if you'd rather we do it in a separate PR.

Regarding naming, a few thoughts:

1) I don't think we should have a method `global_symbol` that returns a `Type`, not a `Symbol`; it should be `global_symbol_ty` instead.
 
2) I'm happy to get rid of the `_by_name` suffix and instead rename e.g. `symbol_ty` to `symbol_ty_by_id` -- by name is the common case and can get the shorter name.

3) I don't have strong feelings either way about `Stdlib::Builtins.global_symbol_ty` vs `builtins_symbol_ty`, but I think I mildly prefer the latter. It's less "clean" in a namespacing sense, but we aren't going to hardcode queries to very many stdlib modules, and `builtins_symbol_ty` is both easier to type and easier to read than `Stdlib::Builtins.global_symbol_ty`, with all its punctuation.

Reply:
(I also prefer `Stdlib` over `CoreStdlibModule`, but that preference is weaker if we provide helper functions like `builtins_symbol_ty` and don't have to use this enum name everywhere.)

Reply:
Well, it will definitely be a simplification, so if you both think it's probably not worth it for these to be Salsa-tracked, I'll defer to you! 🤞 that it doesn't lead to Codspeed complaining 😆